### PR TITLE
splitlist: Add splitadd command.

### PR DESCRIPTION
**Pull request by pandame**
This should fix issue #11.

### CODE REVIEW

File: splitlist/splitlist_commands.c [hunk 1]
```diff
@@ -73,3 +92,81 @@ int spcmd_splitdel(void *source, int cargc, char **cargv) {
...
+    controlreply(np, "Server name %s is too long (max: %d characters)",
+        servername, SERVERLEN);
+    return CMD_ERROR;
+  }
+
+  /* Handle flags */
+  if (cargc > 1) {
+    if (setflags(&servertype, (flag_t)-1, cargv[1], servertypeflags,
+          REJECT_UNKNOWN) != REJECT_NONE) {
+      controlreply(np, "Flag string %s contained invalid flags.\n", cargv[1]);
```
Comment (retropc) [inline splitlist/splitlist_commands.c]: I don't think this should have a newline!

File: splitlist/splitlist_commands.c [hunk 2]
```diff
@@ -1,25 +1,44 @@
...
 MODULE_VERSION("");
 
 int spcmd_splitlist(void *source, int cargc, char **cargv);
 int spcmd_splitdel(void *source, int cargc, char **cargv);
+int spcmd_splitadd(void *source, int cargc, char **cargv);
 
 void _init(void) {
   registercontrolhelpcmd("splitlist", NO_STAFF, 0, &spcmd_splitlist, "Usage: splitlist\nLists servers currently split from the network.");
   registercontrolcmd("splitdel", 10, 1, &spcmd_splitdel);
+  registercontrolhelpcmd("splitadd", 10, 3, &spcmd_splitadd,
```
Comment (retropc) [inline splitlist/splitlist_commands.c]: does the 10 have a constant representation?

File: splitlist/splitlist_commands.c [hunk 3]
```diff
@@ -1,25 +1,44 @@
...
+      " Adds a server as split from the network.\n"
+      " Flags:\n"
+      "  +c: Client server\n"
+      "  +h: Hub server\n"
+      "  +s: Service\n"
+      "  +Q: Q/CServe\n"
+      "  +S: S/spamscan\n"
+      "  +X: Other critical service\n"
+      " If no flags are given, an attempt to figure them out based on name\n"
+      " will be made, but it's likely not a good one.\n");
```
Comment (retropc) [inline splitlist/splitlist_commands.c]: should the last line have a newline?

File: splitlist/splitlist_commands.c [hunk 4]
```diff
@@ -73,3 +92,81 @@ int spcmd_splitdel(void *source, int cargc, char **cargv) {
...
+  servername = cargv[0];
+  servernamelen = strlen(servername);
+
+  if (findserver(servername) != -1) {
+    controlreply(np, "Server %s is linked right now, refusing to add split.",
+        servername);
+    return CMD_ERROR;
+  }
+
+  if (servernamelen > SERVERLEN) {
```
Comment (retropc) [inline splitlist/splitlist_commands.c]: what happens if the server is already in the splitlist? (I'm not sure)

File: splitlist/splitlist_commands.c [hunk 5]
```diff
@@ -73,3 +92,81 @@ int spcmd_splitdel(void *source, int cargc, char **cargv) {
...
+  }
+
+  /* Handle timestamp */
+  if (cargc < 3) {
+    splittime = getnettime();
+  } else {
+    errno = 0;
+    num = strtoull(cargv[2], &end, 10);
+    if (errno == ERANGE) {
+      controlreply(np, "%s is out of range for a timestamp.\n", cargv[2]);
```
Comment (retropc) [inline splitlist/splitlist_commands.c]: \n

File: splitlist/splitlist_commands.c [hunk 6]
```diff
@@ -73,3 +92,81 @@ int spcmd_splitdel(void *source, int cargc, char **cargv) {
...
+    freesstring(fake.name);
+  }
+
+  /* Handle timestamp */
+  if (cargc < 3) {
+    splittime = getnettime();
+  } else {
+    errno = 0;
+    num = strtoull(cargv[2], &end, 10);
+    if (errno == ERANGE) {
```
Comment (retropc) [inline splitlist/splitlist_commands.c]: can this fail any other way?

File: splitlist/splitlist_commands.c [hunk 7]
```diff
@@ -73,3 +92,81 @@ int spcmd_splitdel(void *source, int cargc, char **cargv) {
...
+      return CMD_ERROR;
+    }
+
+    /* Truncation may happen here. 
+     * However, there's no way to get the max time_t value, so we'll just try to
+     * find out after the fact.
+     */
+    splittime = (time_t)num;
+
+    if ((unsigned long long)splittime < num) {
```
Comment (retropc) [inline splitlist/splitlist_commands.c]: wow! I've never seen anyone bother checking this :)

File: splitlist/splitlist_commands.c (hunk 2 above)
Comment (pandame) [inline splitlist/splitlist_commands.c]: @retropc Look at control.h:51: `#define __NO_LEGACY      0x008 /* reserved for old newserv commands, their level is 10 */`

10 gets transformed to flags +LO somewhere down the road. Changing from 10 would mean changing the level to splitadd/splitdel, which I'd need QuakeNet to decide. I don't know enough about your oper access structure to even make a suggestion.

File: splitlist/splitlist_commands.c (hunk 4 above)
Comment (pandame) [inline splitlist/splitlist_commands.c]: It would add a duplicate entry. My bad, I'll fix that.

File: splitlist/splitlist_commands.c (hunk 6 above)
Comment (pandame) [inline splitlist/splitlist_commands.c]: `EINVAL` would signal the last argument, i.e. the base, to strtoull is invalid. While the C standard (seemingly) does not guarantee support for any conversion base, I'd guess encountering `EINVAL` for base 10 is completely and utterly asinine.

The C standard only mentions `ERANGE` as a way to fail, but I can replace it with a generic `errno != 0` check if you'd like me to.

File: splitlist/splitlist_commands.c (hunk 7 above)
Comment (pandame) [inline splitlist/splitlist_commands.c]: So am I supposed to remove it?

**Comment by pandame** [comment]
Added a few comments that require your input. Please look over them.

File: splitlist/splitlist_commands.c (hunk 7 above)
Comment (retropc) [inline splitlist/splitlist_commands.c]: not at all!

File: splitlist/splitlist_commands.c (hunk 6 above)
Comment (retropc) [inline splitlist/splitlist_commands.c]: it's fine as it is :)

File: splitlist/splitlist_commands.c (hunk 2 above)
Comment (retropc) [inline splitlist/splitlist_commands.c]: consistency is good, I'd just leave it then

**Comment by retropc** [comment]
responded, cheers!

**Comment by pandame** [comment]
Hopefully this one sucks less.

**Pull request merged**